README.md
# date_time_picker_form_field
Date and Time picker that uses Futters FormField and Decoration.

## Demo

![](./demo.gif)

## Status

🚧 In Construction 🚧

## Usage
add a line like this to your package's pubspec.yaml (and run an implicit flutter pub get)
```yaml
dependencies:
  date_time_picker_form_field: 
    git:
      url: https://github.com/LeonardoJRibeiro/date_time_picker_form_field
```


## Example

```dart
import 'package:date_time_picker_form_field/date_time_picker_form_field.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(Example());
}

class Example extends StatefulWidget {
  const Example({Key? key}) : super(key: key);
  @override
  _ExampleState createState() => _ExampleState();
}

class _ExampleState extends State<Example> {
  final dateTimeController = DateTimePickerController();
  final formKey = GlobalKey<FormState>();
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Date Timer Picker Form Field Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: Scaffold(
        body: SafeArea(
          child: Form(
            key: formKey,
            child: Column(
              children: [
                DateTimePickerFormField(
                  firstDate: DateTime.now()
                      .subtract(const Duration(days: 356, hours: 6)),
                  lastDate:
                      DateTime.now().add(const Duration(days: 356, hours: 6)),
                  controller: dateTimeController,
                  validator: (date) {
                    if (date == null) {
                      return 'This is a required field!';
                    }
                    if (date.isBefore(DateTime.now())) {
                      return 'Must be now or after than now!';
                    }
                  },
                ),
                Padding(
                  padding: const EdgeInsets.all(8.0),
                  child: ValueListenableBuilder<DateTimePickerValues>(
                    valueListenable: dateTimeController,
                    builder: (context, value, child) => Text(
                        'Date and Time: ${dateTimeController.value.date} | ${dateTimeController.value.time}\nDate and Time in a single object: ${dateTimeController.dateAndTime}'),
                  ),
                ),
                TextButton(
                  onPressed: () {
                    formKey.currentState?.validate();
                  },
                  child: const Text('Validate'),
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }
}
```


## Todo:
-  [ ] accept only date or time on picker, currently it work with both
-  [ ] write documentation
-  [ ] tests




CHANGELOG.md
## 0.0.1 2021-09-19
* initial publish.
## 0.0.2 2021-09-26
* add only date or only time and start of documentation.


lib/widgets/date_time_picker_form_field.dart
part of date_time_picker_form_field;

class DateTimePickerFormField extends FormField<DateTime?> {
  DateTimePickerFormField({
    Key? key,
    FocusNode? focusNode,

    ///The decoration of input
    InputDecoration? decoration,

    ///Is the earliest allowable date.
    required DateTime firstDate,
    //Is the latest allowable date.
    required DateTime lastDate,

    ///Controller thats keep the date or time pickeds, it extends [ValueNotifier] and can be used with a ValueListenableBuild to display the values stored in its state without a [setState].
    required DateTimePickerController controller,

    ///An optional method that validates an input. Returns an error string to display if the input is invalid, or null otherwise.
    String? Function(DateTime? dateTime)? validator,

    ///The initial DateTime.
    DateTime? initialDateTime,
    //The initila Date.
    DateTime? initialDate,

    ///The initial Time
    String? initialTime,

    ///The picker will pick only date wehn true.
    bool onlyDate = false,

    ///The picker will pick only time wehn true.
    bool onlyTime = false,
  })  : assert(!onlyDate || !onlyTime),
        assert(initialTime == null || !onlyDate),
        assert(initialDate == null || !onlyTime),
        assert(initialDateTime == null ||
            !(initialDate != null || initialTime != null)),
        super(
          key: key,
          validator: (_) => validator?.call(controller.dateAndTime),
          builder: (field) {
            return ValueListenableBuilder(
              valueListenable: controller,
              builder: (context, value, child) => DateTimePicker(
                focusNode: focusNode ?? FocusNode(),
                controller: controller,
                firstDate: firstDate,
                lastDate: lastDate,
                decoration: decoration,
                onlyDate: onlyDate,
                onlyTime: onlyTime,
                error: field.errorText,
              ),
            );
          },
        ) {
    if (initialDateTime != null) {
      controller.value = DateTimePickerValues(
        date: DateTime(
          initialDateTime.year,
          initialDateTime.month,
          initialDateTime.day,
        ),
        time: TimeOfDay(
          hour: initialDateTime.hour,
          minute: initialDateTime.minute,
        ),
      );
    } else {
      if (initialDate != null) {
        controller.value = DateTimePickerValues(date: initialDate);
      }
      if (initialTime != null) {
        final date = DateTime.tryParse('0000-00-00T$initialTime');
        if (date != null) {
          controller.value =
              DateTimePickerValues(time: TimeOfDay.fromDateTime(date));
        }
      }
    }
  }
}

class DateTimePicker extends StatefulWidget {
  const DateTimePicker({
    Key? key,
    required this.focusNode,
    this.decoration,
    this.error,
    required this.firstDate,
    required this.lastDate,
    required this.controller,
    required this.onlyDate,
    required this.onlyTime,
  }) : super(key: key);

  final FocusNode focusNode;
  final InputDecoration? decoration;
  final String? error;
  final DateTime firstDate;
  final DateTime lastDate;
  final DateTimePickerController controller;
  final bool onlyDate;
  final bool onlyTime;
  @override
  _DateTimePickerState createState() => _DateTimePickerState();
}

class _DateTimePickerState extends State<DateTimePicker> {
  bool _isHovering = false;

  void _handleFocusChange() {
    setState(() {});
  }

  void _handleHover(bool hover) {
    setState(() {
      _isHovering = hover;
    });
  }

  @override
  void initState() {
    widget.focusNode.addListener(_handleFocusChange);
    super.initState();
  }

  @override
  void dispose() {
    widget.focusNode.removeListener(_handleFocusChange);
    widget.focusNode.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    final localizations = MaterialLocalizations.of(context);
    return InkWell(
      focusNode: widget.focusNode,
      onTap: () async {
        DateTime? date;
        TimeOfDay? time;
        if (!widget.onlyTime) {
          date = await showDatePicker(
            context: context,
            initialDate: widget.controller.dateTime ?? DateTime.now(),
            firstDate: widget.firstDate,
            lastDate: widget.lastDate,
          );
        }
        if (!widget.onlyDate) {
          time = await showTimePicker(
            context: context,
            initialTime: widget.controller.time ?? TimeOfDay.now(),
          );
        }
        if (widget.onlyDate && date != null) {
          widget.controller.value = DateTimePickerValues(
            date: date,
          );
        } else if (widget.onlyTime && time != null) {
          widget.controller.value = DateTimePickerValues(
            time: time,
          );
        } else if (date != null && time != null) {
          widget.controller.value = DateTimePickerValues(
            date: date,
            time: time,
          );
        }
      },
      child: MouseRegion(
        onEnter: (_) => _handleHover(true),
        onExit: (_) => _handleHover(false),
        child: InputDecorator(
          decoration: (widget.decoration ?? const InputDecoration())
              .applyDefaults(Theme.of(context).inputDecorationTheme)
              .copyWith(
                errorText: widget.error,
              ),
          isFocused: widget.focusNode.hasFocus,
          isHovering: _isHovering,
          isEmpty: widget.controller.dateTime == null &&
              widget.controller.time == null,
          child: Text(
              '${widget.controller.dateTime != null ? localizations.formatCompactDate(widget.controller.dateTime!) : ''} ${widget.controller.time != null ? localizations.formatTimeOfDay(widget.controller.time!) : ''}'),
        ),
      ),
    );
  }
}


example/lib/main.dart
import 'package:date_time_picker_form_field/date_time_picker_form_field.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(Example());
}

class Example extends StatefulWidget {
  const Example({Key? key}) : super(key: key);
  @override
  _ExampleState createState() => _ExampleState();
}

class _ExampleState extends State<Example> {
  final dateTimeController = DateTimePickerController();
  final formKey = GlobalKey<FormState>();
  Widget build(BuildContext context) {
    return MaterialApp(
      debugShowCheckedModeBanner: false,
      title: 'Date Timer Picker Form Field Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: Scaffold(
        body: SafeArea(
          child: Form(
            key: formKey,
            child: Column(
              children: [
                DateTimePickerFormField(
                  firstDate: DateTime.now()
                      .subtract(const Duration(days: 356, hours: 6)),
                  lastDate:
                      DateTime.now().add(const Duration(days: 356, hours: 6)),
                  controller: dateTimeController,
                  validator: (date) {
                    if (date == null) {
                      return 'This is a required field!';
                    }
                    if (date.isBefore(DateTime.now())) {
                      return 'Must be now or after than now!';
                    }
                  },
                ),
                Padding(
                  padding: const EdgeInsets.all(8.0),
                  child: ValueListenableBuilder<DateTimePickerValues>(
                    valueListenable: dateTimeController,
                    builder: (context, value, child) => Text(
                        'Date and Time: ${dateTimeController.value.date} | ${dateTimeController.value.time}\nDate and Time in a single object: ${dateTimeController.dateAndTime}'),
                  ),
                ),
                TextButton(
                  onPressed: () {
                    formKey.currentState?.validate();
                  },
                  child: const Text('Validate'),
                ),
              ],
            ),
          ),
        ),
      ),
    );
  }
}
